### PR TITLE
Flexi Raft will considering current config to determine voter list an…

### PR DESCRIPTION
**Pull request by anirbanr-fb**
…d other changes.

Summary:
In Flexi Raft we use "Majority" commit rule along with a voter_distribution provided
to Raft by automation. The voter_distribution and the Majority rule together determine
the number of votes Flexi-Raft will wait for commit in region. So for a region with 1 database + 2 witnesses,
voter_distribution will say {"region1" : 3} and the Majority will mean that LEADER will wait for 2 ACKs.
However if we add 2 more members due to manual mitigations or during  replacements, we need to account
for the increased voter count. Now voter count = 5 and majority should be 3. This was not happening.

The fix is to consider #ACKS_REQUIRED = Majority(MAX(Current Voters, Voter Distribution)).
The voter distribution of 3 helps us in preventing #ACKS_REQUIRED to ever go below 2.
For new region turnups, we want to make sure that the voter distribution map for Raft.

We will disallow any Leader elections in regions where the voter distribution map is not present.
We will also disallow addition of voters without a proper voter distribution map.
This will check that automation like add-member is doing the right thing.
Flexi-Raft for Single Region mode currently operates in 2 ways (controlled by a flag)
1. We elect new LEADER by getting majority from current leader region.
2. We elect new LEADER by getting votes from majority of regions (majority of majority of regions)

Both these approaches had a gap where the majority in the Candidates region was not considered/mandated.
This can cause an availability issue where a single healthy voter in an unhealthy region can
win the election but wouldn't be able to establish authority on the Ring as it would
fail to push the No-Op (the other voters in that region are unhealthy or not present).
This change alters this behavior by waiting for majority from candidate
region.

Test Plan: Unit test to demonstrate the gaps and this patch will fix it,
is to be written on mysql raft side.

Reviewers: iRitwik, bhatvinay, yashtc, yichenshen

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: src/kudu/consensus/quorum_util.h [hunk 1]
```diff
@@ -141,6 +141,19 @@ bool ShouldEvictReplica(const RaftConfigPB& config,
...
+void GetRegionalCountsFromConfig(
+    const RaftConfigPB& config, const std::string& leader_uuid,
+    std::map<std::string, int>* regional_count,
+    std::string* leader_region = nullptr);
+
+// Make each regions voter count, the max of voter distribution and current
+// voters
+void AdjustVoterDistributionWithCurrentVoters(
+    const RaftConfigPB& config,
+    std::map<std::string, int> *voter_distribution);
```
Comment (iRitwik) [inline src/kudu/consensus/quorum_util.h:155]: I don't think we should put in this hack where we calculate a maximum. How about just updating the map within the active config (RaftConfigPB) addition and removal of members? This can be done selectively -- lets say on additions but not on removals or whatever custom logic you want to achieve. Going that route would prevent unnecessary changes to FlexiRaft protocol.

File: src/kudu/consensus/consensus_queue.cc [hunk 2]
```diff
@@ -1159,6 +1159,13 @@ int64_t PeerMessageQueue::ComputeNewWatermark(
       queue_state_.active_config->voter_distribution().begin(),
       queue_state_.active_config->voter_distribution().end());
 
+  // Compute number of voters in each region in the active config.
+  // As voter distribution provided in topology config can lag,
+  // we need to take into account the active voters as well due to
+  // membership changes.
+  AdjustVoterDistributionWithCurrentVoters(
+      *(queue_state_.active_config), &voter_distribution);
+
```
Comment (iRitwik) [inline src/kudu/consensus/consensus_queue.cc:1168]: If we modify the active config's voter distribution, we would not need to mess around with leader election or data commit quorums here. It would be a clean separation between configuration changes and FlexiRaft logic.

File: src/kudu/consensus/leader_election.cc [hunk 3]
```diff
@@ -193,12 +201,27 @@ void FlexibleVoteCounter::FetchTopologyInfo() {
...
+  // As voter distribution provided in topology config can lag,
+  // we need to take into account the active voters as well due to
+  // membership changes.
+  AdjustVoterDistributionWithCurrentVoters(config_, &voter_distribution_);
+
+  // We assume that there are no voters in config_.peers() who
+  // are in present in regions not covered by voter_distribution_
+  // That is enforced via bootstrap and add-member
+  // The reverse is not true and has been handled in
+  // AdjustVoterDistributionWithCurrentVoters
```
Comment (iRitwik) [inline src/kudu/consensus/leader_election.cc:215]: This change would also be unnecessary if we modify the active config directly like I mentioned before.

File: src/kudu/consensus/leader_election.cc [hunk 4]
```diff
@@ -209,31 +232,50 @@ FlexibleVoteCounter::FlexibleVoteCounter(
...
   for (const std::pair<std::string, int>& regional_voter_count :
       voter_distribution_) {
-    CHECK_GT(regional_voter_count.second, 0);
-    num_voters_ += regional_voter_count.second;
+    // When instances are being removed from ring, the voter distribution
+    // can have extra regions, but we have taken them out in
+    // FetchTopology. So this should never happen
+    if (regional_voter_count.second <= 0) {
+      continue;
+    }
```
Comment (iRitwik) [inline src/kudu/consensus/leader_election.cc:240]: Shouldn't we do these validations during config changes?

File: src/kudu/consensus/leader_election.cc [hunk 5]
```diff
@@ -245,6 +287,8 @@ Status FlexibleVoteCounter::RegisterVote(
       no_vote_count_[region]++;
       break;
   }
+
+  // TODO - explain this more
```
Comment (iRitwik) [inline src/kudu/consensus/leader_election.cc:291]: stale comments?

File: src/kudu/consensus/leader_election.cc [hunk 6]
```diff
@@ -424,6 +468,15 @@ std::pair<bool, bool> FlexibleVoteCounter::IsStaticQuorumSatisfied() const {
...
 
+// Flexible Paxos/Raft says that for a quorum of N if Datapath quorum is
+// M then leader election quorum is (N+1) - M
+// In the case of region based quorums, if M = 1 (single region dynamic)
+// The leader election quorum should include all regions.
+// In the case that all regions are healthy, this pessimistic quorum would
+// suffice. We use this first because this is the most comprehensive check,
+// however in case regions are down, this would not work. In those cases
+// we use other heuristics like intersecting with last leader region or
+// a majority of regions + last leader region.
```
Comment (iRitwik) [inline src/kudu/consensus/leader_election.cc:479]: nit: move this to header?

File: src/kudu/consensus/leader_election.cc [hunk 7]
```diff
@@ -943,6 +1001,8 @@ std::pair<bool, bool> FlexibleVoteCounter::IsDynamicQuorumSatisfied() const {
     }
   }
 
+  // This could be empty as a Last known leader could no more
+  // be a voter or in the ring.
```
Comment (iRitwik) [inline src/kudu/consensus/leader_election.cc:1005]: or during bootstrap or upon being explicitly cleared.

File: src/kudu/consensus/leader_election.cc [hunk 8]
```diff
@@ -436,7 +489,7 @@ FlexibleVoteCounter::IsPessimisticQuorumSatisfied() const {
   }
 
   // Getting quorum satisfaction for individual regions.
-  const std::vector<std::pair<bool, bool> >& results =
+  std::vector<std::pair<bool, bool> > results =
```
Comment (iRitwik) [inline src/kudu/consensus/leader_election.cc:492]: why is this required?

File: src/kudu/consensus/leader_election.cc [hunk 9]
```diff
@@ -958,30 +1018,77 @@ std::pair<bool, bool> FlexibleVoteCounter::IsDynamicQuorumSatisfied() const {
...
+  // only starts elections in voters which have valid region
+  std::string candidate_region(
+      DetermineRegionForUUID(candidate_uuid_));
+
   // Step 3: Check if last known leader's quorum is satisfied and we directly
   // succeed term.
-  if (election_term_ == last_known_leader.election_term() + 1) {
+  // Since pessimistic quorum is not satisfied, we have to intersect with last
+  // known leader region to guarantee longest log in next LEADER.
+  // However if there is a period of confusion after there was a stable LEADER,
```
Comment (iRitwik) [inline src/kudu/consensus/leader_election.cc:1030]: what is a "period of confusion"? :)

File: src/kudu/consensus/leader_election.cc [hunk 10]
```diff
@@ -958,30 +1018,77 @@ std::pair<bool, bool> FlexibleVoteCounter::IsDynamicQuorumSatisfied() const {
...
+  // the CANDIDATE might not be able to know which region to intersect with.
+  //
+  // If a CANDIDATE has heard from all voters then it should have confidence
+  // that its crowdsourced last known leader is the correct one. So if there
+  // is no term continuity, it only means that there was no leader during the
+  // period of confusion
+  bool continuity_not_required =
+      FLAGS_crowdsource_last_known_leader && FLAGS_trust_last_leader_entries &&
+      AreAllVotesIn() && !last_known_leader.uuid().empty();
```
Comment (iRitwik) [inline src/kudu/consensus/leader_election.cc:1039]: I am not sure this is safe. Why was this done?

File: src/kudu/consensus/leader_election.cc [hunk 11]
```diff
@@ -958,30 +1018,77 @@ std::pair<bool, bool> FlexibleVoteCounter::IsDynamicQuorumSatisfied() const {
...
+  //
   // Step 4: Find possible leader regions at every term greater than last
   // known leader's term. Computes possible successor regions until next term
   // is the current election's term or the quorum converges to pessimistic
   // quorum.
+  //
+  // TODO (anirbanr-fb) - This needs to be evaluated from the standpoint
+  // of success depending on votes from both leader region and candidate
+  // region. Postponing since it is relatively complicated logic here, and
+  // the current angle of attack might not work.
```
Comment (iRitwik) [inline src/kudu/consensus/leader_election.cc:1091]: Not sure what these comments mean. What is the problem we are trying to solve?

File: src/kudu/consensus/raft_consensus.cc [hunk 12]
```diff
@@ -556,12 +556,18 @@ Status RaftConsensus::StartElection(ElectionMode mode, ElectionReason reason) {
       return Status::IllegalState("only voting members can start elections",
           SecureShortDebugString(cmeta_->ActiveConfig()));
     }
-    if (PREDICT_FALSE(active_role == RaftPeerPB::NON_PARTICIPANT)) {
-      SnoozeFailureDetector();
-      return Status::IllegalState("Not starting election: node is currently "
-                                  "a non-participant in the Raft config",
-                                  SecureShortDebugString(cmeta_->ActiveConfig()));
```
Comment (iRitwik) [inline src/kudu/consensus/raft_consensus.cc:563]: Why was this removed? Can non-participants now start elections?

File: src/kudu/consensus/raft_consensus.cc [hunk 13]
```diff
@@ -556,12 +556,18 @@ Status RaftConsensus::StartElection(ElectionMode mode, ElectionReason reason) {
...
+
+    // In flexi raft mode, we want to start elections only in Candidate
+    // regions which have voter_distribution Information.
+    if (FLAGS_enable_flexi_raft) {
+      const auto& vd_map = cmeta_->ActiveConfig().voter_distribution();
+      if (PREDICT_FALSE(vd_map.find(peer_region()) == vd_map.end())) {
+        return Status::IllegalState(strings::Substitute(
+            "in flexi-raft only regions with valid voter distribution can start election: $0",
+            peer_region()));
+      }
```
Comment (iRitwik) [inline src/kudu/consensus/raft_consensus.cc:568]: Shouldn't this validation live where we do config validations -- during bootstrap and / or when members are added or removed?

File: src/kudu/consensus/leader_election.h [hunk 14]
```diff
@@ -130,6 +130,7 @@ class VoteCounter {
 class FlexibleVoteCounter : public VoteCounter {
  public:
   FlexibleVoteCounter(
+      const std::string& candidate_uuid,
```
Comment (iRitwik) [inline src/kudu/consensus/leader_election.h:133]: This will break all the FlexibleVoteCounter tests but should be easy to fix. I am not yet convinced of the utility of mandating a majority in own region. Reasons here: https://fburl.com/svuygk2q

If we must do this, may I suggest reusing the `PreviousVotePB` protobuf to pass in the candidate_uuid and election term as one param? We can rename `PreviousVotePB` to `CandidateInfoPB` or something.

File: src/kudu/consensus/quorum_util.h (hunk 1 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/quorum_util.h:155]: There is no change in flexi-raft protocol for this. It modifies voter_distribution which actually should have been
named target_voters/minimum_voters to begin with. FlexiRaft's majority calculation has not changed.

File: src/kudu/consensus/consensus_queue.cc (hunk 2 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/consensus_queue.cc:1168]: We should keep voter distribution in the config based on what automation passed us. We don't want to second guess and lose track of what is user provided.

File: src/kudu/consensus/leader_election.cc (hunk 4 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/leader_election.cc:240]: Yes we should and we do it as well. However there is no point crashing the server for something
that we can handle.

File: src/kudu/consensus/leader_election.cc (hunk 5 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/leader_election.cc:291]: We can explain what we do with uuid_to_last_term_pruned_ in more detail.

File: src/kudu/consensus/leader_election.cc (hunk 8 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/leader_election.cc:492]: because const reference here does not make sense. You are getting a return by value.

File: src/kudu/consensus/leader_election.cc (hunk 9 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/leader_election.cc:1030]: period of confusion is my way of saying that there was no leader in that term.

File: src/kudu/consensus/leader_election.cc (hunk 11 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/leader_election.cc:1091]: Which part don't you understand?
I tried to fit "get Votes from candidate Region" logic as well into ComputeElectionResultFromVotingHistory and it goes region by region, instead of term by term. I feel ComputeElectionResultFromVotingHistory is more complicated or atleast needs to be explained a bit more here.

File: src/kudu/consensus/leader_election.cc (hunk 10 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/leader_election.cc:1039]: This is an optimization to not have to go to ComputeElectionResultFromVotingHistory.
When we have all votes from all voters, we should know who the last leader is because
the voters have sent it to us. term == lkl_term + 1 is restrictive. 

If we have lkl in term 2 but a failed election in term 3, 4 , 5  and candidate starting election in term 6 and all voters have told us that there is no known leader beyond term 2, we can assume that there is no known leader && if we have the requisite votes from them. Do you have a counter example?

File: src/kudu/consensus/quorum_util.h (hunk 1 above)
Comment (iRitwik) [reply src/kudu/consensus/quorum_util.h:155]: I am suggesting to not change this within FlexiRaft code at all. When config changes are done --  addition / removal of members from the replicaset, why not modify the map there? we are modifying active config there already.
https://github.com/anirbanr-fb/kuduraft/blob/1.8.raft/src/kudu/consensus/metadata.proto#L239

If we do that, majority of this diff is not required.

File: src/kudu/consensus/leader_election.cc (hunk 4 above)
Comment (iRitwik) [reply src/kudu/consensus/leader_election.cc:240]: Not asking to crash -- just reject the config changes if validation fails.

File: src/kudu/consensus/consensus_queue.cc (hunk 2 above)
Comment (iRitwik) [reply src/kudu/consensus/consensus_queue.cc:1168]: Not sure I follow the logic here, isn't what is user provided already present in ProdConfig? If you are worried about automation making a mistake, we should add logging to automation.

File: src/kudu/consensus/leader_election.cc (hunk 8 above)
Comment (iRitwik) [reply src/kudu/consensus/leader_election.cc:492]: sure. but why remove const-ness?

File: src/kudu/consensus/leader_election.cc (hunk 10 above)
Comment (iRitwik) [reply src/kudu/consensus/leader_election.cc:1039]: The whole point of `ComputeElectionResultFromVotingHistory` is to figure out that there is no leader beyond term 2. Why is this change required?

File: src/kudu/consensus/raft_consensus.cc (hunk 12 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/raft_consensus.cc:563]: the check before this check prevents non-participants to start elections. its a redundant check

File: src/kudu/consensus/quorum_util.h (hunk 1 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/quorum_util.h:155]: I don't think we should modify consensus metadata to keep it up to date with config. Firstly because it is not the same as voter list. When we remove voters from config we would not change it to go below 3. So it will only be done on addition and not on all removals. So it is already highly heursitical.

It will become a nightmare to debug why we are out of synch with what was provided to us by automation. 
If you want to add another field called target_voters_distribution_ and put it in metadata.proto then it would make sense, but then we would have to modify FlexiRaft to move to the new field.

**Review by yashtc — APPROVED** [review]
Accepting to unblock RPM build

File: src/kudu/consensus/leader_election.h (hunk 14 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/leader_election.h:133]: I will setup a discussion to discuss this within the few of us. We can discuss this idea there.
